### PR TITLE
WWST-1019 Add IKEA TRADFRI RGB bulbs

### PR DESCRIPTION
**Pull request by dkirker**
This is a work in progress. There are still a few rough edges as this is the first implementation of CIE 1931 xyY color.

### CODE REVIEW

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy [hunk 1]
```diff
@@ -73,14 +98,33 @@ def parse(String description) {
...
+			}
+			else if(zigbeeMap.attrInt == ATTRIBUTE_Y) { //Y Attribute
+				def yValue = zigbee.convertHexToInt(zigbeeMap.value)
+				log.debug "yValue = $yValue"
+
+				state.currentY = yValue
+			}
+
+			if (!shouldUseHueSaturation() && state.currentX && state.currentY) {
+				def hsv = safeColorXy2Hsv(state.currentX, state.currentY)
```
Comment (varzac) [inline devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy]: Did you do testing of this on the cloud side.  I was working on this locally and the hue/sat values that we get out don't really match up with the colors.

**Comment by varzac** [comment]
Tom and I were talking and we've decided that we are OK with just updating the min hubcore version to 25 to handle the issue of people joining on pre 25 hubs.

**Comment by dkirker** [comment]
Thanks! I will update this PR.

**Review by tpmanley — DISMISSED** [review]

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /**
- *  Copyright 2017 SmartThings
+ *  Copyright 2017-2019 SmartThings
```
Comment (tpmanley) [inline devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy]: Best not to put an end date. Without it the copyright continues to the current date (I think)

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy [hunk 3]
```diff
@@ -67,27 +83,46 @@ def parse(String description) {
...
 				def hueValue = Math.round(zigbee.convertHexToInt(zigbeeMap.value) / 0xfe * 100)
 				sendEvent(name: "hue", value: hueValue, displayed:false)
-			}
-			else if(zigbeeMap.attrInt == ATTRIBUTE_SATURATION){ //Saturation Attribute
+			} else if(zigbeeMap.attrInt == ATTRIBUTE_SATURATION && shouldUseHueSaturation()) { //Saturation Attribute
 				def saturationValue = Math.round(zigbee.convertHexToInt(zigbeeMap.value) / 0xfe * 100)
 				sendEvent(name: "saturation", value: saturationValue, displayed:false)
+			} else if(zigbeeMap.attrInt == ATTRIBUTE_X) { //X Attribute
+				state.currentRawX = zigbee.convertHexToInt(zigbeeMap.value)
+				//log.debug "xValue = $state.currentRawX"
```
Comment (tpmanley) [inline devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy]: Should remove commented out code

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy [hunk 4]
```diff
@@ -67,27 +83,46 @@ def parse(String description) {
...
 		sendEvent(finalResult)
-	}
-	else {
+	} else {
 		def zigbeeMap = zigbee.parseDescriptionAsMap(description)
 		log.trace "zigbeeMap : $zigbeeMap"
 
 		if (zigbeeMap?.clusterInt == COLOR_CONTROL_CLUSTER) {
-			if(zigbeeMap.attrInt == ATTRIBUTE_HUE){  //Hue Attribute
+			if(zigbeeMap.attrInt == ATTRIBUTE_HUE && shouldUseHueSaturation()){ //Hue Attribute
```
Comment (tpmanley) [inline devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy]: Nits: space after ifs

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy [hunk 5]
```diff
@@ -67,27 +83,46 @@ def parse(String description) {
...
+			} else if(zigbeeMap.attrInt == ATTRIBUTE_X) { //X Attribute
+				state.currentRawX = zigbee.convertHexToInt(zigbeeMap.value)
+				//log.debug "xValue = $state.currentRawX"
+			} else if(zigbeeMap.attrInt == ATTRIBUTE_Y) { //Y Attribute
+				state.currentRawY = zigbee.convertHexToInt(zigbeeMap.value)
+				//log.debug "yValue = $state.currentRawY"
 			}
-		}
-		else {
+
```
Comment (tpmanley) [inline devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy:102]: Nit: extra line

**Review by tpmanley — DISMISSED** [review]
Just a few minor comments

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy (hunk 2 above)
Comment (dkirker) [reply devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy]: Hmmmm Maybe that is what I had heard. Though, a quick look seems to indicate that really none of it matters? But, I'll revert that commit, and fix the couple of other code formatting things that I missed.

File: devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy (hunk 5 above)
Comment (dkirker) [reply devicetypes/smartthings/zll-rgb-bulb.src/zll-rgb-bulb.groovy:102]: I think diff view made this look weird. :/

**Review by tpmanley — APPROVED** [review]

**Pull request merged**